### ISSUE TITLE
Enhance TLS with a CA: attribute

### ISSUE THREAD

**Issue by SittingDuc** [Koenkk/zigbee2mqtt#523]
Feature request.

What I am wanting to do on my local hass.io, is to move one step past self-signed certificates. So I have created my own "root-ca.pem" and signed my certificate with it. But I cannot see any way to feed this knowledge into zigbee2mqtt.

Reading up on MQTT and TLS and node.js, The Internet states node.js compiles the list of CA Certificates into the binary; which is a blocker; but also that the tls.connect() function accepts a 'ca' option which would let me add (open question: append or overstrike?) a local certificate.

So it stands to reason a new feature could be added to allow the user to set a CA Certificate in the existing configuration.yaml and add it to the mqtt.connect() options field, so it will be passed to tls.connect() and everything will just work.

Now, I know nothing about javascript, node.js, or github. And I am running hass.io, so even if I could work out how to edit a local fork, I can't compile it / test it locally. Which means I am not well placed to help. But I tried anyway!

So, I have written some changes to lib/mqtt.js. How would I best get my change to y'all for review, fixing and merging?

I've tried attaching a patch/diff file to this statement. Maybe that will work?

[patch_mqtt_js_tlsca.txt](https://github.com/Koenkk/zigbee2mqtt/files/2500109/patch_mqtt_js_tlsca.txt)

TL;DR:
1. here is a patch against lib/mqtt.js that reads a certfile from disk, named by the user in configuration.yaml, and adds it to the mqtt.connect() option field and thence to the tls.connect() option field
2. I don't know how to javascript, I can't test the change, I bet it has a spelling mistake!
3. I don't know if the ca option field overstrikes the original CA list, or appends the new certificate :(
4. I don't know enough about github or this project to work out how to git push the change "properly"
5. If it is of any use, you are welcome to it!

**Comment by tb-killa** [comment]
Hi thanks for your Patch.

From what i read you would use your own ca and cert file in combination with ``mqtts``.

@Koenkk 
We need only:
--mqtts as server-port (could be defined by settings.yaml)
--``fs.readFileSync`` with user specified path for (need validation too):
---client.key
---client.crt
---ca.crt ??
and pass this as options:
--ca: CAfile,
--key: KEY,
--cert: CERT,


@SittingDuc Does i miss something ??

**Comment by SittingDuc** [comment]
Howdy,

If we have an unauthenticated client connecting to an encrypted server (my use-case), the client doesn't need Key or Cert, it just needs some way to validate the server's presented certificate matches the server - i.e checking the CA-chain.
Since I am self-signed in my use-case, I am presenting the CA-chain for my own root and my own intermediate to zigbee2mqtt, with the intent that if the server's presented cert is signed by one of those, the connection is allowed. (and if anyone MITM my intranet, the authentication fails)

A separate topic is if we add support for client certificates, so the server can authenticate the client is approved/allowed. This isn't my use-case (I feel passwords are sufficient for my intranet). In this case, the client gets a Key and a Cert, and keeps the CA-chain to authenticate the server. (and the server has it's own CA-chain to authenticate the client).

We probably need some clear unambiguous way to separate the two use-cases. I had to read the docs on mqtt bridges several times and find an example before I understood which tokens applied to the upstream connection and which tokens applied to the downstream connection.
